presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Rule.java

```diff
@@ -34,7 +34,18 @@ default boolean isEnabled(Session session)
         return true;
     }
 
-    Optional<PlanNode> apply(T node, Captures captures, Context context);
+    @Deprecated
+    default Optional<PlanNode> apply(T node, Captures captures, Context context)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    // TODO remove default implementaiton once deprecated apply method got removed
+    default void apply(T node, Captures captures, Context context, Result result)
```

Comment:
The most practical usage I was thinking was enumerating `MultiJoinNode`. Memo could decide to stop accepting new results if `JoinEnumerationRule` produces a lot of results, but I guess we can add such interface later on and wrap result returning method.

> This would be only possible if (a) you use interrupts and trust them to be effective

In a single code base you cannot trust other components to yield. It's up to them if they do independently of the mechanism. With iterator they might produce full results first anyway. It doesn't mean that you don't want yielding to happen (see: https://github.com/prestodb/presto/pull/8785)